### ISSUE TITLE
persist: more introspection for `persist_source`

### ISSUE THREAD

**Issue by aljoscha** [MaterializeInc/materialize#16392]
We cannot currently look into the inner workings of `persist_source` (which itself is a dataflow fragment consisting of three operators) as it is running, which makes debugging incidents in production hard. We should at least add metrics that allow us to look at:
 - number of discovered batch parts
 - number of fetched batch parts
 - number of fetched and decoded updates
 - (maybe) number of times we yielded
 - capability held at the batch discovery operator
 - capability held at the fetcher operator

[1]: internal thread on an incident: https://materializeinc.slack.com/archives/C04CKCRSUDC/p1669755204457019

**Comment by danhhz** [comment]
> number of fetched batch parts

I _think_ we have this one already

> capability held at the ...

how are you thinking we model this as a metric?

**Comment by aljoscha** [comment]
> how are you thinking we model this as a metric?

in storage metrics we assume that time is totally ordered and just report the first entry of the `CapabilitySet` that these operators holds as a metric. It allows us to (approximately) see the output frontier at the different operators that make up the source.

**Comment by danhhz** [comment]
oh, sorry, I was asking more how we'd aggregate them. like max/min of every persist_source on the process? group/label by source_id and do max/min? group by individual persist_source instance somehow? something something by worker?

**Comment by aljoscha** [comment]
> oh, sorry, I was asking more how we'd aggregate them. like max/min of every persist_source on the process? group/label by source_id and do max/min? group by individual persist_source instance somehow? something something by worker?

yes, those are the good questions! I think to be really useful it would be good to have them "by source_id, by worker", but that blows them up a bit. Plus, the persist metrics so far don't know about source ids or workers, so it feels these metrics might want to "live" on the Storage metrics dashboard? And maybe have the `mz_storage` prefix instead of `mz_persist`, wdyt? Either would be fine for me, though.

**Comment by danhhz** [comment]
yeah no strong feels on which dashboard they're on or which prefix. I'm skeptical that prom metrics with `source_id, worker` labels are worth the cost though. kinda feels like we have a hammer and are looking for nails

**Comment by danhhz** [comment]
don't we get this sort of frontier information from the introspection queries, anyway?

**Comment by aljoscha** [comment]
Closing this for now, there was no more interest (also from my side), and the persist team will add any required metrics as more features are added.
